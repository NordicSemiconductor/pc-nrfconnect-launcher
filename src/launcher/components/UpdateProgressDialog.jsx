/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import ProgressBar from 'react-bootstrap/ProgressBar';
import { Spinner } from 'pc-nrfconnect-shared';
import { bool, func, number, string } from 'prop-types';

const UpdateProgressDialog = ({
    isVisible,
    isProgressSupported,
    isCancelSupported,
    version,
    percentDownloaded,
    onCancel,
    isCancelling,
}) => (
    <Modal show={isVisible} backdrop>
        <Modal.Header closeButton={false}>
            <Modal.Title>Downloading update</Modal.Title>
        </Modal.Header>
        <Modal.Body>
            <p>Downloading nRF Connect {version}...</p>
            {isProgressSupported && (
                <ProgressBar
                    label={`${percentDownloaded}%`}
                    now={percentDownloaded}
                />
            )}
            <p>
                This might take a few minutes. The application will restart and
                update once the download is complete.
            </p>
        </Modal.Body>
        <Modal.Footer>
            {!isProgressSupported && <Spinner />}
            {isCancelSupported && (
                <Button
                    onClick={onCancel}
                    disabled={isCancelling || percentDownloaded === 100}
                >
                    Cancel
                </Button>
            )}
        </Modal.Footer>
    </Modal>
);

UpdateProgressDialog.propTypes = {
    isVisible: bool.isRequired,
    isProgressSupported: bool.isRequired,
    isCancelSupported: bool.isRequired,
    version: string.isRequired,
    percentDownloaded: number.isRequired,
    onCancel: func.isRequired,
    isCancelling: bool.isRequired,
};

export default UpdateProgressDialog;
