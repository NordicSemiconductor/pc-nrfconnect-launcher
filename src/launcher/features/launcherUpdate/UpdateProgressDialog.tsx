/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import ProgressBar from 'react-bootstrap/ProgressBar';
import { Dialog, DialogButton, Spinner } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { cancelDownload } from './launcherUpdateEffects';
import { getLauncherUpdate } from './launcherUpdateSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const {
        isUpdateProgressDialogVisible: isVisible,
        isProgressSupported,
        isCancelSupported,
        latestVersion: version,
        percentDownloaded,
        isCancelling,
    } = useLauncherSelector(getLauncherUpdate);

    return (
        <Dialog isVisible={isVisible} closeOnUnfocus={false}>
            <Dialog.Header title="Downloading update" />
            <Dialog.Body>
                <p>Downloading nRF Connect for Desktop {version}...</p>
                {isProgressSupported && (
                    <ProgressBar
                        label={`${percentDownloaded}%`}
                        now={percentDownloaded}
                    />
                )}
                <p>
                    This might take a few minutes. The application will restart
                    and update once the download is complete.
                </p>
            </Dialog.Body>
            <Dialog.Footer>
                {!isProgressSupported && <Spinner />}
                {isCancelSupported && (
                    <DialogButton
                        onClick={() => dispatch(cancelDownload())}
                        disabled={isCancelling || percentDownloaded === 100}
                        variant="primary"
                    >
                        Cancel
                    </DialogButton>
                )}
            </Dialog.Footer>
        </Dialog>
    );
};
