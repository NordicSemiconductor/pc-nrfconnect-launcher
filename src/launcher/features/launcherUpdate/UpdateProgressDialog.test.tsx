/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import render from '../../../test/testrenderer';
import {
    cancelDownload,
    startDownload,
    updateAvailable,
    updateDownloading,
} from './launcherUpdateSlice';
import UpdateProgressDialog from './UpdateProgressDialog';

describe('UpdateProgressDialog', () => {
    it('is initially invisible', () => {
        expect(render(<UpdateProgressDialog />).baseElement).toMatchSnapshot();
    });

    it('can have a version, percent downloaded, and be cancellable', () => {
        expect(
            render(<UpdateProgressDialog />, [
                updateAvailable('1.2.3'),
                startDownload({
                    isCancelSupported: true,
                    isProgressSupported: true,
                }),
                updateDownloading(42),
            ]).baseElement,
        ).toMatchSnapshot();
    });

    it('can be without progress, and not cancellable', () => {
        expect(
            render(<UpdateProgressDialog />, [
                updateAvailable('1.2.3'),
                startDownload({
                    isCancelSupported: false,
                    isProgressSupported: false,
                }),
                updateDownloading(42),
            ]).baseElement,
        ).toMatchSnapshot();
    });

    it('has cancelling disabled after being cancelled', () => {
        expect(
            render(<UpdateProgressDialog />, [
                updateAvailable('1.2.3'),
                startDownload({
                    isCancelSupported: true,
                    isProgressSupported: true,
                }),
                cancelDownload(),
            ]).baseElement,
        ).toMatchSnapshot();
    });

    it('has cancelling disabled when 100 percent complete', () => {
        expect(
            render(<UpdateProgressDialog />, [
                updateAvailable('1.2.3'),
                startDownload({
                    isCancelSupported: true,
                    isProgressSupported: true,
                }),
                updateDownloading(100),
            ]).baseElement,
        ).toMatchSnapshot();
    });
});
