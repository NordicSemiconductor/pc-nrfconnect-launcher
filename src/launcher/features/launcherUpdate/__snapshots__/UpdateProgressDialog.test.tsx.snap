// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProgressDialog can be without progress, and not cancellable 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade dialog  modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div>
            <b>
              Downloading update
            </b>
            <div
              class="spinner-border spinner-border-sm"
            />
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            Downloading nRF Connect for Desktop 
            1.2.3
            ...
          </p>
          <p>
            This might take a few minutes. The application will restart and update once the download is complete.
          </p>
        </div>
        <div
          class="modal-footer"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`UpdateProgressDialog can have a version, percent downloaded, and be cancellable 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade dialog  modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div>
            <b>
              Downloading update
            </b>
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            Downloading nRF Connect for Desktop 
            1.2.3
            ...
          </p>
          <div
            class="progress"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="42"
              class="progress-bar"
              role="progressbar"
              style="width: 42%;"
            >
              42%
            </div>
          </div>
          <p>
            This might take a few minutes. The application will restart and update once the download is complete.
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class=""
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`UpdateProgressDialog has cancelling disabled after being cancelled 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade dialog  modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div>
            <b>
              Downloading update
            </b>
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            Downloading nRF Connect for Desktop 
            1.2.3
            ...
          </p>
          <div
            class="progress"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0"
              class="progress-bar"
              role="progressbar"
              style="width: 0%;"
            >
              0%
            </div>
          </div>
          <p>
            This might take a few minutes. The application will restart and update once the download is complete.
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class=""
            disabled=""
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`UpdateProgressDialog has cancelling disabled when 100 percent complete 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade dialog  modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div>
            <b>
              Downloading update
            </b>
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            Downloading nRF Connect for Desktop 
            1.2.3
            ...
          </p>
          <div
            class="progress"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="100"
              class="progress-bar"
              role="progressbar"
              style="width: 100%;"
            >
              100%
            </div>
          </div>
          <p>
            This might take a few minutes. The application will restart and update once the download is complete.
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class=""
            disabled=""
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`UpdateProgressDialog is initially invisible 1`] = `
<body>
  <div />
</body>
`;
